### PR TITLE
feature: Implement SwaggerUI api key authentication functionality

### PR DESCRIPTION
**Pull request by bartwesselink**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | api-platform/docs#232

This PR integrates api key functionality for Swagger UI into api-platform.

### CODE REVIEW

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 1]
```diff
@@ -546,19 +550,36 @@ private function computeDoc(Documentation $documentation, \ArrayObject $definiti
...
+                'scopes' => $this->oauthScopes,
+            ];
+
+            $security[] = ['oauth' => []];
+        }
+
+        if ($this->apiKeyEnabled) {
+            $securityDefinitions['apiKey'] = [
+                'type' => 'apiKey',
+                'in' => 'header',
```
Comment (meyerbaptiste) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: What about the `query` value?

> Specify whether the API key will be passed in: header or in: query.

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 2]
```diff
@@ -546,19 +550,36 @@ private function computeDoc(Documentation $documentation, \ArrayObject $definiti
...
+                'description' => sprintf('Value for the %s header', $this->apiKeyHeader),
+                'name' => $this->apiKeyHeader,
             ];
 
-            $doc['security'] = [['oauth' => []]];
+            $security[] = ['apiKey' => []];
+        }
+
+        if (count($securityDefinitions) > 0 && count($security) > 0) {
```
Comment (meyerbaptiste) [inline src/Swagger/Serializer/DocumentationNormalizer.php]:
```php
if ($securityDefinitions && $security) {
```

**Comment by bartwesselink** [comment]
@meyerbaptiste thanks for reviewing my PR! I didn't even know it was possible to add more api keys to Swagger UI. I've updated my PR with the changes you suggested. PR for docs has also been updated.

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 3]
```diff
@@ -546,19 +548,41 @@ private function computeDoc(Documentation $documentation, \ArrayObject $definiti
...
+                $type = $apiKey['type'];
+
+                $securityDefinitions[$key] = [
+                    'type' => 'apiKey',
+                    'in' => $type,
+                    'description' => sprintf('Value for the %s %s', $name, $type === 'query' ? sprintf('%s parameter', $type) : $type),
+                    'name' => $name,
+                ];
+
+                $security[] = [$key => []];
```
Comment (meyerbaptiste) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: This only allows a logical OR but AND support can be added in another PR.

e.g.:
```
// OR
$security = [];
$security[] = ['foo' => []];
$security[] = ['bar' => []];

// AND
$security = [];
$security[0] = ['foo' => []];
$security[0] += ['bar' => []];
```

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php [hunk 4]
```diff
@@ -219,6 +219,11 @@ public function testEnableNelmioApiDoc()
         $this->extension->load(array_merge_recursive(self::DEFAULT_CONFIG, ['api_platform' => ['enable_nelmio_api_doc' => true]]), $containerBuilder);
     }
 
+    public function testEnableApiKey()
+    {
+        $this->assertTrue(true);
```
Comment (meyerbaptiste) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php]: Is this a Volkswagen test? 🤣

**Comment by bartwesselink** [comment]
Ah sorry! Going to close this because I don't have time to create an updated version (lots of merge conflicts)

**Comment by soyuka** [comment]
No problem, I'll keep this open maybe someone will continue the work on this feature :). Thanks for the work so far!

**Comment by Simperfit** [comment]
on it ;)

**Comment by Simperfit** [comment]
ping @meyerbaptiste @soyuka @bartwesselink

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -86,6 +86,20 @@ public function getConfigTreeBuilder()
                     ->end()
                 ->end()
 
+                ->arrayNode('api_keys')
```
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Maybe we should put oauth and this under a `swagger` node. IMO the name is too generic.

Same with the parameter, I think `api_platform.swagger.api_keys` is cleaner then `api_platform.api_keys` which is misleading.

Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Yeah I totally agree

Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Changing oauth ins't a BC break?

Comment (meyerbaptiste) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Nop, oauth was introduced in the master branch.

Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: changer oauth isn't a BC break?

**Comment by Simperfit** [comment]
ping @soyuka done, I think oAuth should be done in another PR that I can do to follow up if we want too

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php (hunk 5 above)
Comment (soyuka) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: > Nop, oauth was introduced in the master branch.
Exactly, it's not in stable yet. Could be done in another PR.

**Comment by soyuka** [comment]
Meh, in fact it could be:

```
swagger:
    enabled: true
    ui: true
    auth:
      oauth: 
     api_key:
```

Sadly we can't break on `swagger_enabled` and `swagger_ui`.

Leave it like this for now! @teohhanhui opinion on this?

**Comment by Simperfit** [comment]
ping @meyerbaptiste

**Comment by soyuka** [comment]
And please squash haha

**Pull request merged**